flink-runtime/src/test/java/org/apache/flink/runtime/metrics/util/MetricUtilsTest.java

```diff
@@ -175,24 +183,20 @@ public MetricGroup addGroup(String name) {
         @SuppressWarnings("unchecked")
         final Gauge<Long> used = (Gauge<Long>) metaspaceMetrics.get(MetricNames.MEMORY_USED);
 
-        final long usedMetaspaceInitially = used.getValue();
+        runUntilMetricChanged("Metaspace", 10, () -> redefineAsNewClass(MetricUtils.class), used);
+    }
 
-        // check memory usage difference multiple times since other tests may affect memory usage as
-        // well
-        for (int x = 0; x < 10; x++) {
-            List<Runnable> consumerList = new ArrayList<>();
-            for (int i = 0; i < 10; i++) {
-                consumerList.add(() -> {});
-            }
+    @Test
+    public void testNonHeapMetricUsageNotStatic() throws Exception {
+        final InterceptingOperatorMetricGroup nonHeapMetrics =
+                new InterceptingOperatorMetricGroup();
 
-            final long usedMetaspaceAfterAllocation = used.getValue();
+        MetricUtils.instantiateNonHeapMemoryMetrics(nonHeapMetrics);
 
-            if (usedMetaspaceInitially != usedMetaspaceAfterAllocation) {
-                return;
-            }
-            Thread.sleep(50);
-        }
-        Assert.fail("Metaspace usage metric never changed it's value.");
+        @SuppressWarnings("unchecked")
+        final Gauge<Long> used = (Gauge<Long>) nonHeapMetrics.get(MetricNames.MEMORY_USED);
+
+        runUntilMetricChanged("Non-heap", 10, () -> redefineAsNewClass(MetricUtils.class), used);
```

Comment:
Could we not define an anonymous class each time?

Reply:
@zentol Did you means named method and method reference ?

Reply:
@zentol Sorry for not get what you means, it is pure due to my english skill. I actually tried, but anonymous class is not resistant to possible optimization. That said it is easily to fail in IDEA "Until Failure".

Reply:
@zentol `testMetaspaceMetricUsageNotStatic` was changed for exactly that reason. Old approach could not last 10 runs.

Reply:
have you checked whether your approach of storing them in `referencedObjects` would do the trick?

Reply:
Hi @zentol, I runs following paths:
* With `referencedObjects` and new anonymous class, it does not last 2 runs in my IDEA.
* Without `referencedObjects` and new anonymous class, same as above.
* With `referencedObjects` and `testNonHeapMetricUsageNotStatic`, 100K runs without failure.
* Without `referencedObjects` and `testNonHeapMetricUsageNotStatic`, 100K runs without failure.

`referencedObjects` does not contribute to result for both approaches. I keep it mainly for mental health that the created object should live before checking. Before `referencedObjects` approach, I actually use a simple assertion statement for this as an replacement of Java 9 `Reference.reachabilityFence`.

Here is code snippet for anonymous class:
```java
        runUntilMetricChanged(
                "Non-heap",
                10,
                () ->
                        new Runnable() {
                            @Override
                            public void run() {
                                System.out.println(used);
                            }
                        },
                used);
```